### PR TITLE
[Metrics] Add metrics to upload operation

### PR DESCRIPTION
**Pull request by ncooke3**
Metrics are added to an upload batch during the upload process.

- This PR integrates the metrics controller into the upload process.
- Updated the `GDTCCTUploaderTest.m` tests to expect metrics in the upload batches

### CODE REVIEW

**Comment by ncooke3** [comment]
Integration tests are failing as metrics controller is still not implemented.

**Comment by ncooke3** [comment]
I'm going to break this PR up for easier review. It contains some code that isn't relevant to the title.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m [hunk 1]
```diff
@@ -140,110 +147,132 @@ - (void)uploadTarget:(GDTCORTarget)target withConditions:(GDTCORUploadConditions
...
+                  return [FBLPromise resolvedWith:batch];
+                }
+
+                return [self batchByAddingMetricsEventToBatch:batch forTarget:target];
+              })
       .validateOn(self.uploaderQueue,
-                  ^BOOL(GDTCORUploadBatch *batch) {
-                    // 5. Validate batch.
-                    return batch.batchID != nil && batch.events.count > 0;
+                  ^BOOL(id result) {
```
Comment (paulb777) [inline GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m:192]: Any way to get better type safety here than `id`?

File: GoogleDataTransport/GDTCORLibrary/GDTCORMetricsController.m [hunk 2]
```diff
@@ -52,30 +52,48 @@ - (instancetype)init {
...
 }
 
-- (void)storage:(nonnull id<GDTCORStoragePromiseProtocol>)storage
-    didDropEvent:(nonnull GDTCOREvent *)event {
-  // TODO(ncooke3): Implement.
+- (BOOL)isMetricsCollectionSupportedForTarget:(GDTCORTarget)target {
+  switch (target) {
+    // Only the Firelog backend supports metrics collection.
+    case kGDTCORTargetFLL:
+    case kGDTCORTargetCSH:
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/GDTCORMetricsController.m:76]: What's CSH?

**Review by paulb777 — APPROVED** [review]
Didn't go too deep, but happy to GVC if you want to do so for any of it.

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m (hunk 1 above)
Comment (ncooke3) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m]: Done in [5324cea](https://github.com/google/GoogleDataTransport/pull/52/commits/5324cea052b93da71cc87b497fff065a840c7543). Discussed offline.

File: GoogleDataTransport/GDTCORLibrary/GDTCORMetricsController.m (hunk 2 above)
Comment (ncooke3) [reply GoogleDataTransport/GDTCORLibrary/GDTCORMetricsController.m:76]: Discussed offline.

**Review by chliangGoogle — APPROVED** [review]

**Pull request merged**